bodhi-server/bodhi/server/models.py

```diff
@@ -2250,16 +2277,27 @@ def _greenwave_api_url(self):
 
     @property
     def _greenwave_decision_contexts(self):
-        # We retrieve updates going to testing (status=pending) and updates
-        # (status=testing) going to stable.
-        # We also query on different contexts for critpath and non-critpath
-        # updates.
+        """
+        Return greenwave contexts that should be queried for this update.
+
+        The base context depends whether the update is going to
+        testing (status=pending) or to stable (status=testing). We
+        also query on additional contexts for critpath updates. If the
+        critpath type provides grouped information, we query on one
+        additional context per critpath group represented in the
+        update's packages. If it does not, we query on the catch-all
+        additional 'critpath' context.
+        """
         # this is correct if update is already in testing...
         contexts = ["bodhi_update_push_stable"]
         if self.request == UpdateRequest.testing and self.status == UpdateStatus.pending:
             # ...but if it is pending, we want to know if it can go to testing
             contexts = ["bodhi_update_push_testing"]
-        if self.critpath:
+        if self.critpath_groups is not None:
```

Comment:
I think this should just be `if self.critpath_groups:` to exclude empty field

Reply:
ah, yeah. my intent there was that we should always hit the first part of the conditional in grouped critpath 'mode', and if the update was not in any critpath groups, it shouldn't do anything, because the string is empty. But I forgot that even splitting an empty string gives you a single-item list.

The way to 'fix' it the way I intended would be this:

        if self.critpath_groups is not None:
            if self.critpath_groups:
                ocontext = contexts[0]
                for group in self.critpath_groups.split(" "):
                    contexts.insert(0, f"{ocontext}_{group}_critpath")
        elif self.critpath:
            contexts.insert(0, contexts[0] + "_critpath")
        return contexts

but your way is less obtuse and should always achieve the same effect, it just means we end up evaluating one extra statement when we're in grouped 'mode' and the update is not in the critpath, which isn't worth bothering about. So I'll go with your approach. I'll have to see if I can make the tests cover this case too.